platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java

```diff
@@ -188,17 +196,35 @@ public void signUriString(String queryParams, UriBuilder uriBuilder) throws Sign
     return signature;
   }
 
-  private String getSignatureAlgorithm(X509Certificate certificate) {
-    String sigAlgo = SSOConstants.RSA_SHA1;
-    String pubKeyAlgo = certificate.getPublicKey().getAlgorithm();
+  private java.security.Signature getSignature(X509Certificate certificate)
+      throws SignatureException {
+    String jceSigAlgo = certificate.getSigAlgName();
```

Comment:
❗️ `getSigAlgName()` returns the signature algorithm that the CA used to sign this certificate. Need to call `certificate.getPublicKey().getAlgorithm()` here (which will just return the encrypting algorithm without the hashing algorithm i.e. "RSA") . And then we need to append the hashing algorithm that we want to use for the signing. `"SHA1with<encrypting-algo>`. NOTE: This should only be used for signing. For validating signatures, we can't use this method because we need to pull the hashing algorithm AND encrypting algorithm from the signature we're given.

Reply:
I want to be sure I've got this right. @brjeter, @blen-desta, and @stustison, is this correct now to the best of your knowledge, where I've replaced the `getSignature` method with two methods, one for signing and one for verification?
```java
  private java.security.Signature getValidationSignature(X509Certificate certificate)
      throws SignatureException {
    String jceSigAlgo = certificate.getSigAlgName();

    java.security.Signature signature;
    try {
      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
        signature = java.security.Signature.getInstance(jceSigAlgo, "BC");
      } else {
        signature = java.security.Signature.getInstance(jceSigAlgo);
      }
    } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
      throw new SignatureException(e);
    }
    return signature;
  }

  private java.security.Signature getSigningSignature(X509Certificate certificate)
      throws SignatureException {
    java.security.Signature signature;
    try {
      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
        signature = java.security.Signature.getInstance("SHA384withDSA", "BC");
      } else {
        signature = java.security.Signature.getInstance("SHA384withRSA");
      }
    } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
      throw new SignatureException(e);
    }
    return signature;
  }
```

Reply:
The `getValidationSignature` method won't be necessary. When validating a signature we need to be given a signature algorithm rather than pulling it off the cert. For Redirect, it's given in the queryParams (see my second comment). For POST, it's already in the OpenSAML `Signature` object that's passed in.

Reply:
What @brjeter said and also we are probably fine with just using the SHA256 signatures